README.md
#Gitso

Gitso is a frontend to reverse VNC connections. It is meant to be a simple two-step process that connects one person to another's screen.

Original Project is kept at: https://code.google.com/p/gitso/

Few changes that I've made:

+ Minimization now minimizes to the task tray
+ When using the command line switch --listen, Gitso starts minimized
+ Can now specify which port to listen on. --listen {PORT}

##Building Gitso
####Windows
In the src folder you will find instructions on how to build Gitso on Windows along with installation files.
**You may have to change some paths in *makegitso.bat*** depending on your installation.

####Mac
These instructions have been taken from https://code.google.com/p/gitso/wiki/BuildOSX

1. Install Developer Tools (Xcode) from the OS X System CD
2. Install py2app
3. From the command line type:
4. curl -O http://peak.telecommunity.com/dist/ez_setup.py
5. sudo python ez_setup.py -U setuptools
6. sudo easy_install -U py2app
7. Update hosts.txt to have preset options for the client. Hosts are comma separated and optional.
8. Run: ./makegitso.sh

####Linux
These instructions have been taken from https://code.google.com/p/gitso/wiki/BuildLinux

**Non Ubuntu, Fedora, and OpenSUSE Users**
 
Gitso requires the following files so make sure to install them:

+ x11vnc
+ vncviewer
+ wxPython

**Fedora Users**

1. yum install subversion
2. yum install rpmdevtools

**Ubuntu, Fedora, OpenSUSE Users**

1. Update hosts.txt to have preset options for the client. Hosts are comma separated and optional.
2. Run: ./makegitso.sh [--source, --opensuse, --fedora]

##Disclaimer
I really have no desire to maintain this project, as it's not mine to begin with.
I was looking for something like Gitso but it didn't quite have what I wanted.
After making my changes I thought I might as well put this up on GitHub for others who wanted
something similar. So if you have issues, you're better off forking the project and fixing
them yourself.


debian/changelog
gitso (0.6.3) rawr; urgency=super high

  * Updated --listen command line to accept port (defaults to 5500)
	* Minimizing now minimizes to tray
	* If --listen is given at start, Gitso will start minimized
	
	-- AustP Thu, 22 May 2014 16:27:00 -0700

gitso (0.6.2) precise; urgency=low

  * Cleaned up .deb creation.
  * Remove CHANGELOG.txt.
  * Make gitso executable.
  * Added homepage to .deb control file.
  * Changed vncviewer dependency to xtightvncviewer.
  * Fix package name in linux changelog.
  * Added 8 bit color compression.

 -- Markus Roth <markus.roth@herr-biber.de>  Thu, 27 Sep 2012 03:06:54 +0200

gitso (0.6) karmic; urgency=low

  * Complete rewrite of process management.
  * Actually stop VNC Processes (Windows)
  * Support loading remote hosts file.
  * Command line switches
  *     --dev
  *     --listen
  *     --connect IP
  *     --list list_file
  *     --version
  *     --help
  * manpage for (All UNIX sytems)
  * Support for .rpms (Fedora, OpenSUSE)
  * Implement Native VNC listener (OS X)
  * Better process management, user gets notified if connection is broken.
  * Licensing Updates (across the board).
  * Improved documentation. 

 -- Aaron D. Gerber <gerberad@gmail.com>  Sun, 21 Feb 2010 17:32:40 -0600

gitso (0.5) hardy; urgency=low

  * Complete rewrite of the interface
  * Gitso no longer has Zombied VNC processes after it quits.
  * Gitso stops the VNC process when it closes
  * Updated Icon
  * Updated License: GPL 3
  * Added Support to be able to specify a list of hosts when you distribute it.
  * Added History/Clear History of servers
  * Added OS X 10.5 Support (need testing on 10.4 and 10.3)
  *   OS X uses TightVNC 1.3.9
  *   OS X uses OSXVNC 3.0

 -- Aaron D. Gerber <gerberad@gmail.com>  Sat, 26 Jul 2008 16:32:40 -0600

gitso (0.4) hardy; urgency=low

  * Made Deb, updated with icons and much more.

 -- Aaron D. Gerber <gerberad@gmail.com>  Sat, 10 May 2008 16:17:43 -0600

gitso (0.3) UNRELEASED; urgency=low

  * Initial release. (Closes: #XXXXXX)

 -- Aaron D. Gerber <gerberad@gmail.com>  Thu, 08 May 2008 22:35:52 -0600


ArgsParser.py
#! /usr/bin/env python

"""
Gisto - Gitso is to support others

Gitso is a utility to facilitate the connection of VNC

@author: Aaron Gerber ('gerberad') <gerberad@gmail.com>
@author: Derek Buranen ('burner') <derek@buranen.info>
@author: AustP
@copyright: 2008 - 2014

Gitso is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Gitso is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Gitso.  If not, see <http://www.gnu.org/licenses/>.
"""

import os
import sys
import signal
import os.path
import urllib
import re

class ArgsParser:
	def __init__(self):
		# Initialize Self.paths here.
		self.paths = dict()
		self.paths['resources'] = os.path.join(sys.path[0], "./")
		self.paths['preferences'] = ''
		self.paths['copyright'] = ''
		self.paths['main'] = ''
		self.paths['listen'] = False
		self.paths['connect'] = ''
		self.paths['list'] = []
		self.paths['mode'] = ''
		self.paths['low-colors'] = False
		self.port = "5500"
		
		if re.match('(?:open|free|net)bsd|linux',sys.platform):
			self.paths['main'] = os.path.join(sys.path[0], '..', 'share', 'gitso')
			self.paths['copyright'] = os.path.join(sys.path[0], '..', 'share', 'doc', 'gitso', 'COPYING')
		elif sys.platform == "darwin":
			self.paths['main'] = sys.path[0]
			self.paths['copyright'] = os.path.join(sys.path[0], 'COPYING')
		else:
			self.paths['main'] = os.path.join(sys.path[0], '..')
			self.paths['copyright'] = os.path.join(sys.path[0], '..', 'COPYING')
		
		#for i in range(1, len(sys.argv)):
		i = 1
		while i < len(sys.argv):
			if sys.argv[i] == '--help': # --help
				self.HelpMenu()
			elif sys.argv[i] == '--version': # --version
				print "Gitso 0.6.3  -- Copyright 2007 - 2014 Aaron Gerber and Derek Buranen and AustP."
				exit(0)
			elif sys.argv[i] == '--dev': # --dev
				print "Running in 'Development Mode'"
				self.paths['mode'] = 'dev'
				if sys.platform == "darwin":
					if not os.path.exists('build/OSXvnc'):
						os.popen("mkdir build; cp arch/osx/OSXvnc.tar.gz build ; cd build ; tar xvfz OSXvnc.tar.gz > /dev/null")
					if not os.path.exists('build/cotvnc.app'):
						os.popen("cp arch/osx/cotvnc.app.tar.gz build ; cd build ; tar xvfz cotvnc.app.tar.gz > /dev/null")
						
					self.paths['resources'] = 'build/'
					self.paths['main']		= sys.path[0]
					self.paths['copyright'] = os.path.join(sys.path[0], 'COPYING')
					
				elif sys.platform == "win32":
					self.paths['copyright'] = os.path.join(sys.path[0], 'COPYING')
					self.paths['main']		= os.path.join(sys.path[0])
					self.paths['resources'] = 'arch/win32/'
					
				else:
					self.paths['resources'] = 'arch/linux/'
					self.paths['main']		= os.path.join(sys.path[0])
					self.paths['copyright'] = os.path.join(sys.path[0], 'COPYING')

			elif sys.argv[i] == '--listen': # --listen
				if self.paths['connect'] <> "":
					print "Error: --connect and --listen can not be used at the same time."
					self.HelpMenu()

				i = i + 1
				if i >= len(sys.argv):
					self.port = "5500"
				else:
					if sys.argv[i][0] + sys.argv[i][1] <> "--":
						self.port = sys.argv[i]
					else:
						self.port = "5500"
						i = i - 1

				self.paths['listen'] = True

			elif sys.argv[i] == '--connect': # --connect
				i = i + 1
				if i >= len(sys.argv):
					print "Error: No IP or domain name given."
					self.HelpMenu()

				if self.paths['listen']:
					print "Error: --connect and --listen can not be used at the same time."
					self.HelpMenu()
				
				if sys.argv[i][0] + sys.argv[i][1] <> "--":
					self.paths['connect'] = sys.argv[i]
				else:
					print "Error: '" + sys.argv[i] + "' is not a valid host with '--connect'."
					self.HelpMenu()

			elif sys.argv[i] == '--low-colors': # --low-colors
				self.paths['low-colors'] = True;

			elif sys.argv[i] == '--list': # --list
				i = i + 1
				if i >= len(sys.argv):
					print "Error: No List file given."
					self.HelpMenu()
				
				if sys.argv[i][0] + sys.argv[i][1] <> "--":
					self.paths['list'] = self.getHosts(sys.argv[i])
				else:
					print "Error: '" + sys.argv[i] + "' is not a valid list with '--list'."
					self.HelpMenu()

			else:
				print "Error: '" + sys.argv[i] + "' is not a valid argument."
				self.HelpMenu()

			i = i + 1
		
		if sys.platform == "darwin":
				self.paths['preferences'] = os.path.join(os.path.expanduser("~"), "Library", "Application Support", "Gitso")
				if os.path.exists(self.paths['preferences']) != True:
						os.makedirs(self.paths['preferences'], 0700)
				self.paths['preferences'] = os.path.join(self.paths['preferences'], "hosts")
		elif sys.platform == "win32":
				self.paths['preferences'] = os.path.join(os.getenv('USERPROFILE'), "gitso-hosts")
		else:
				self.paths['preferences'] = os.path.join(os.path.expanduser("~"), ".gitso-hosts")

	#Help Menu
	def HelpMenu(self):
		print "Usage: " + os.path.basename(sys.argv[0]) + " [OPTION]"
		print "   OPTIONS"
		print "   --dev\t\tSet self.paths for development."
		print "   --listen {PORT}\tListen for incoming connections."
		print "   --connect {IP|DN}\tConnects to host (support giver)."
		print "   --list {URL|FILE}\tAlternative Support list."
		print "   --low-colors\t\tUse 8bit colors (for slow connections). Linux only."
		print "   --version\t\tThe current Gitso version."
		print "   --help\t\tThis Menu."
		sys.exit(1)
	
	def GetPaths(self):
		return self.paths
		
	def GetPort(self):
		return self.port
		
	def getHosts(self, file):
		list = []
		fileList = ""
		
		if len(file) > 3:
			prefix = file[0] + file[1] + file[2] + file[3]
		else:
			prefix = ""
		
		if prefix == "www." or prefix == "http":
			handle = urllib.urlopen(file)
			fileList = handle.read()
			handle.close()
		else:
			if os.path.exists(file):
				handle = open(file, 'r')
				fileList = handle.read()
				handle.close()
		
		parsedlist = fileList.split(",")
		for i in range(0, len(parsedlist)):
			if self.validHost(parsedlist[i].strip()):
				list.append(parsedlist[i].strip())
		
		return list

	def validHost(self, host):
		if host != "" and host.find(";") == -1 and host.find("/") == -1 and host.find("'") == -1 and host.find("`") == -1 and len(host) > 6:
			return True
		else:
			return False

